### DISCUSSION TITLE
New Demo: Standalone Link Routing with Libavoid

### DISCUSSION THREAD

**Discussion by kumilingus** [clientIO/joint#2627] [News]
[Libavoid](https://www.adaptagrams.org/documentation/libavoid.html) is a C++ library that offers high-quality polyline and **orthogonal link routing**, effectively **navigating around objects**, and **avoiding overlapping parallel links**.  It is particularly useful when used in conjunction with interactive diagram editors such as those that can be created with *JointJS*.

Today's [example](https://changelog.jointjs.com/rappid/examples/libavoid/) demonstrates how to use this library (specifically the [libavoid-js](https://github.com/Aksem/libavoid-js) lib which is a port to JavaScript) to **automatically route the links** in a *JointJS* diagram.

![Image](https://github.com/orgs/clientIO/projects/7/assets/3967880/a45e3fe1-56dc-475d-9b5d-ef56094d675f)

The source code is available on [GitHub](https://github.com/clientIO/joint/tree/master/examples/libavoid).

**Comment by shailendra-kbtechnology** [comment]
Nice implementation 
have you code in type script ?

**Reply by kumilingus** [reply]
No. The example is available in *JavaScript* only.

**Reply by shailendra-kbtechnology** [reply]
Alright,
i tried the implement in to the ts.
But we are gating error

![image](https://github.com/clientIO/joint/assets/156991175/da07f2b7-7c2a-4033-af03-39d2c4a82e87)

**Reply by kumilingus** [reply]
I wonder if you are missing this [step](https://github.com/clientIO/joint/blob/master/examples/libavoid/webpack.config.js#L40-L45).

**Reply by shailendra-kbtechnology** [reply]
We saw the code but our project in next and we are not using webpack file,
So there have any alternate option?

**Reply by kumilingus** [reply]
Yeah, you just need to copy the `./node_modules/libavoid-js/dist/libavoid.wasm` to the folder the application is being served from.

**Comment by shailendra-kbtechnology** [comment]
Thanks @kumilingus

**Reply by csruchitaoza** [reply]
Hey, are you able to archive that library in TypeScript? If yes then could you please share knowledge or perhaps a small demo?

I tried to convert it to JSON but faced the same error as you did.
I'm using angular 17.

Thank you,

**Comment by trufak** [comment]
Hi!
Please tell me if you have ever encountered such an error:
"RuntimeError: Aborted(CompileError: WebAssembly.instantiate(): expected magic word 00 61 73 6d, found 3c 21 44 4f @+0)"
It looks like the wasm file is not binary

**Reply by kumilingus** [reply]
See https://github.com/WebAssembly/spec/issues/1410#issuecomment-1014597652.

**Comment by trufak** [comment]
@kumilingus, is it possible to implement the exclusion of some elements from the number of obstacles ?

**Reply by kumilingus** [reply]
It should be easy if the elements themselves are not connected to any links.
Just return early [here](https://github.com/clientIO/joint/blob/master/examples/libavoid/src/avoid-router.js#L159) if the `element` is not an obstacle.

**Reply by trufak** [reply]
The element represents a container that has output ports. Lines connect the output ports of the container to its child elements. In this case, the vertices of the lines are built trying to go around the parent container. 
![image](https://github.com/user-attachments/assets/38404b9e-8a95-4b1e-98fb-6683ec154717)

How can I make sure that the parent container is not an obstacle for child lines?

**Reply by kumilingus** [reply]
Then I don't know.

You can try asking the author of [libavoid](https://github.com/mjwybrow/adaptagrams) if that is supported or not. The other question is whether this functionality has been ported to the [JavaScript version](https://github.com/Aksem/libavoid-js) or not.

**Comment by bernarde01** [comment]
This is a great feature! When will this be included in a release version? I am running core 4.0.4, and getPortGroupNames is not yet in this version (although I can see it was added in June to the git repo)

**Reply by kumilingus** [reply]
Hi, in November. In the mean time, you can replace:

`getPortGroupNames()`

with:

`Object.keys(element.prop('ports/groups'));`.

**Reply by marty0678** [reply]
Hey thanks for this demo! I tried to get it working a little while ago but didn't have much luck, but I didn't spend much time on it. I need to take another crack at it soon - but saw this comment, will this be part of the official JointJS / JointJS+ library at some point based on this comment? That would be awesome if so! The Manhattan router is great but it starts to break down at medium-large drawings.

Thanks again for this awesome library!

**Reply by kumilingus** [reply]
Right, we have plans to provide a wrapper for the `libavoid` 3rd party library as a standalone package e.g. `@joint/libavoid`.

If you want to start using it now, all you need to do is [instantiate](https://github.com/clientIO/joint/blob/master/examples/libavoid/src/ui-thread/app.js#L230-L237) the [`AvoidRouter` class](https://github.com/clientIO/joint/blob/master/examples/libavoid/src/shared/avoid-router.js#L6).

**Reply by marty0678** [reply]
Just realized I never responded to this, sorry. That's awesome to hear about the future plans. And the issues I had when trying to instantiate it were with the WASM loading, more so than the Joint side, I think. I still haven't had a chance to get back to this, but hopefully I will soon! Thank you for the response!
